hoodie-common/src/main/java/com/uber/hoodie/common/util/buffer/MemoryBoundedBuffer.java

```diff
@@ -151,47 +190,20 @@ private void readNextRecord() {
       }
     }
     if (newRecord.isPresent()) {
-      this.nextRecord = newRecord.get();
+      return newRecord;
     } else {
       // We are done reading all the records from internal iterator.
       this.isDone.set(true);
-      this.nextRecord = null;
-    }
-  }
-
-  public void startBuffering() throws Exception {
-    logger.info("starting to buffer records");
-    try {
-      while (inputIterator.hasNext()) {
-        // We need to stop buffering if buffer-reader has failed and exited.
-        throwExceptionIfFailed();
-        insertRecord(inputIterator.next());
-      }
-      // done buffering records notifying buffer-reader.
-      buffer.put(Optional.empty());
-    } catch (Exception e) {
-      logger.error("error buffering records", e);
-      // Used for notifying buffer-reader thread of the failed operation.
-      markAsFailed(e);
-      throw e;
-    }
-    logger.info("finished buffering records");
-  }
-
-  @Override
-  public boolean hasNext() {
-    if (this.nextRecord == null && !this.isDone.get()) {
-      readNextRecord();
+      return Optional.empty();
     }
-    return !this.isDone.get();
   }
 
-  @Override
-  public O next() {
-    Preconditions.checkState(hasNext() && this.nextRecord != null);
-    final O ret = this.nextRecord;
-    this.nextRecord = null;
-    return ret;
+  /**
+   * Puts an empty entry to buffer to denote termination
+   */
+  public void stopProducing() throws InterruptedException {
```

Comment:
rename:  closeBuffer() ?

Also do we save the fact that this method was called to prevent more insertRecord() calls being done? Would be good to do that